File: cpp/0446.0_Arithmetic_Slices_II_-_Subsequence.cpp
/*
DP
T: O(N^2)
S: O(N^2)

Runtime: 1153 ms, faster than 29.59% of C++ online submissions for Arithmetic Slices II - Subsequence.
Memory Usage: 135.4 MB, less than 69.15% of C++ online submissions for Arithmetic Slices II - Subsequence.
*/

#define LL long long
class Solution {
public:
    int numberOfArithmeticSlices(vector<int>& nums) {
        int n = nums.size();
        LL ans = 0;
        vector<map<LL, int>> dp(n);
        for (int i = 1; i < n; i++) {
            for (int j = 0; j < i; j++) {
                LL delta = (LL)nums[i] - (LL)nums[j];
                int cnt = 0;
                if (dp[j].find(delta) != dp[j].end()) {
                    cnt = dp[j][delta];
                }
                dp[i][delta] += cnt + 1;
                ans += cnt;
            }
        }
        
        return (int)ans;
    }
};

File: problems/0446.0_Arithmetic_Slices_II_-_Subsequence.py
'''
DP + hash table
T: O(N^2)
S: O(N^2)

Runtime: 3064 ms, faster than 29.00% of Python3 online submissions for Arithmetic Slices II - Subsequence.
Memory Usage: 52.4 MB, less than 68.64% of Python3 online submissions for Arithmetic Slices II - Subsequence.
'''
class Solution:
    def numberOfArithmeticSlices(self, nums: List[int]) -> int:
        n = len(nums)
        ans = 0
        dp = [Counter() for _ in range(n)]
        # i -> j
        for j in range(1, n):
            for i in range(j):
                diff = nums[j] - nums[i]
                ans += dp[i][diff]
                dp[j][diff] += dp[i][diff] + 1
        return ans
        
